data/src/main/java/com/linkedin/data/ByteString.java

```diff
@@ -1584,4 +1618,42 @@ private int getCurrentRemaining()
       return _currentSegment.getOffset() + _currentSegment.getLength() - _localOffset;
     }
   }
+
+  private static class ByteStringLongDecoderState extends Utf8Utils.LongDecoderState
+  {
+    private final ByteArrayVector _byteArrays;
+
+    private int _currentIndex;
+
+    ByteStringLongDecoderState(ByteArrayVector byteArrays, int currentIndex, int localOffset)
```

Comment:
> currentIndex is index to the array of ByteArray
Yes

> localOffset is the index to current ByteArray
Yes, this is the current position in the byte array

ByteArray.offset represents the offset of the underlying byte[] of the ByteArray. This is needed to know when the byte array ends, since we operate only on raw byte[] in Utf8Utils.

Reply:
+1 on this renaming.

Reply:
But sorry, I still didn't quite get the difference btw current `ByteArray.offset` and this `localOffset`. This `localOffset` is also index to the underlying byte[] of  current ByteArray, no?

Reply:
offset is the offset of the underlying byte[] when creating the ByteArray. Let us say its is 4.
localOffset represents current position in the byte[]. Let us say it is 8.